src/utils/buffers/histogram_buffer.rs

```diff
@@ -0,0 +1,367 @@
+            let mid = left + (right - left) / 2;
+            if value >= self.edges[mid] {
+                left = mid + 1;
+            } else {
+                right = mid - 1;
+            }
+        }
+        // Return the bin index
+        left as usize
+    }
+}
+
+impl VisualizerBuffer<f32> for HistogramBuffer {
+    fn enqueue(self: &mut Self, value: f32) {
+        let value = value.abs();
+        // don't enqueue silence
+        if value > 0.0 {
+            let bin_index = self.find_bin(value);
+            self.data[bin_index] += (1.0 - self.decay_weight); // Increment the count for the bin
+            for i in 0..self.size - 1 {
+                // decay all values
+                self.data[i] *= self.decay_weight;
+            }
+        }
+    }
+
+    fn enqueue_buffer(
+        self: &mut Self,
+        buffer: &mut nih_plug::buffer::Buffer,
+        channel: Option<usize>,
+    ) {
+        match channel {
+            Some(channel) => {
+                for sample in buffer.as_slice()[channel].into_iter() {
+                    self.enqueue(*sample);
+                }
+            }
+            None => {
+                for sample in buffer.iter_samples() {
+                    self.enqueue(
+                        (1. / (&sample).len() as f32) * sample.into_iter().map(|x| *x).sum::<f32>(),
+                    );
+                }
+            }
+        }
+    }
+
+    /// Resizes the buffer to the given size, **clearing it**.
+    fn resize(self: &mut Self, size: usize) {
+        if size == self.len() {
+            return;
+        }
+        self.clear();
+        self.size = size;
+        self.update();
+    }
+
+    /// Clears the entire buffer, filling it with default values (usually 0)
+    fn clear(self: &mut Self) {
+        self.data.iter_mut().for_each(|x| *x = f32::default());
+    }
+
+    fn len(self: &Self) -> usize {
+        self.size
+    }
+
+    /// Grows the buffer, **clearing it**.
+    fn grow(self: &mut Self, size: usize) {
+        self.resize(size);
+    }
+
+    /// Shrinks the buffer, **clearing it**.
+    fn shrink(self: &mut Self, size: usize) {
+        self.resize(size)
+    }
+}
+
+impl Index<usize> for HistogramBuffer {
+    type Output = f32;
+
+    fn index(&self, index: usize) -> &Self::Output {
+        if index >= self.size {
+            panic!(
+                "Invalid histogram buffer access: Index {} is out of range for histogram buffer of size {}",
+                index, self.size
+            );
+        }
+        &self.data[index]
+    }
+}
+impl IndexMut<usize> for HistogramBuffer {
+    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
+        if index >= self.size {
+            panic!(
+                "Invalid histogram buffer access: Index {} is out of range for histogram buffer of size {}",
+                index, self.size
+                );
+        }
+        &mut self.data[index]
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
These tests all seem to fail. Do you want to fix them? Testing the buffers is usually a bit too tricky to be worth it, as most of them do floating-point math that can be kind of a pain to test; You might just be better off removing them.